crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -375,6 +375,24 @@ pub struct ReadOnlyUserIdentity {
     user_id: OwnedUserId,
     pub(crate) master_key: Arc<MasterPubkey>,
     self_signing_key: Arc<SelfSigningPubkey>,
+    /// The first time an identity is seen for a given user it will be marked as
+    /// tofu trusted. If a new identity is detected for that user, we must
+    /// ensure that this change has been shown and validated by the user. If
+    /// this boolean is set to `false`, sharing room keys to this user might
+    /// fail depending on the room key sharing strategy.
+    #[serde(
+        default = "tofu_trusted_default",
+        serialize_with = "atomic_bool_serializer",
+        deserialize_with = "atomic_bool_deserializer"
+    )]
+    tofu_trusted: Arc<AtomicBool>,
```

Comment:
`tofu_trusted` doesn't seem like a great name for for a flag that means *either* "the first identity we saw for this user" *or* "not the first identity, but the user acknowledged it". TBH, I'm wondering if we should actually split it into two different flags.

Reply:
Or an enum, given that the two flags would be mutually exclusive?

Reply:
> Or an enum, given that the two flags would be mutually exclusive?

I did think about that, but we still have the problem of what to call the damn thing

Reply:
I discussed this with @BillCarsonFr and semantics of the flag are roughly whether the identity is pinned or not, so perhaps the name could be `pinned`.

However, the way it's currently implemented—where we immediately replace the stored identity as soon as a new one is returned from the server, but with this flag set to `false`—means that a server can delete your pin by simply presenting a new identity and then switching back to the old one, resulting in the flag switching from `true` to `false`.

Valere is going to experiment with an alternative approach, where we

- Remember the old pinned identity as long as a new identity doesn't become pinned, replacing the old one
- Using the logic `current != pinned` to control whether the user should be notified / prompted to confirm

Reply:
@dkasak Alternative for MSK pinning https://github.com/matrix-org/matrix-rust-sdk/pull/3639